### PR TITLE
Feature rest tutorial

### PR DESCRIPTION
**Pull request by Plesoun**
@ateska Please see and comment. Its .md but I think I can convert it to the .rst

### CODE REVIEW

File: doc/tutorial/web/rest/asab.webservice.md [hunk 1]
```diff
@@ -0,0 +1,271 @@
+Creating REST HTTP webservice using  ASAB
+===
+
+## Overview
+
+On lines below, we will discuss a creation of a basic ASAB microservice application capable of 
+providing a REST interface. We will be using MongoDB as an example database we are trying to
```
Comment (ateska) [inline doc/tutorial/web/rest/asab.webservice.md:7]: providing a REST HTTP API.

File: doc/tutorial/web/rest/asab.webservice.md [hunk 2]
```diff
@@ -0,0 +1,271 @@
+Creating REST HTTP webservice using  ASAB
+===
+
+## Overview
+
+On lines below, we will discuss a creation of a basic ASAB microservice application capable of 
+providing a REST interface. We will be using MongoDB as an example database we are trying to 
+perform the Create Read Update and Delete operations on.
```
Comment (ateska) [inline doc/tutorial/web/rest/asab.webservice.md:8]: "we are trying to perform" -> "we perform" :-)

File: doc/tutorial/web/rest/asab.webservice.md [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+2. Asynchronous Server App Boilerplate (ASAB) version 20.3 or later
+3. MongoDB instance
+4. Postman
+
+## Components
+
+Our sample microservice will consist of several modules, each one serving a logical purpose within the
+whole. The components are as follows (and they also hint a filestructure), and will be discussed in greater 
+detail in their respective sections below, we will go from top to bottom:   
+`myrestapi.py`, `./myrestapi/app.py`, `./myrestapi/crud/handler.py`, and `./myrestapi/crud/service.py`
```
Comment (ateska) [inline doc/tutorial/web/rest/asab.webservice.md:22]: `crud` folder should be named rather as `tutorial` eg. - not by a technology.

File: doc/tutorial/web/rest/asab.webservice.md [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+on how to quickly do that.   
+We will be using Postman to test the API and to generate json collection of available endpoints.
+
+#### myrestapi.py
+
+This is the part, from where we run our app. We begin with the shebang line, that will tell the 
+executing operating system, that we want python to execute this program.  
+Linux: `#!/usr/bin/python3`  
+Windows `#! python3`   
+OS X `#!/usr/bin/env python3`
```
Comment (ateska) [inline doc/tutorial/web/rest/asab.webservice.md:33]: The standard is `!/usr/bin/env python3` ... and it should be agnostic to OS.

File: doc/tutorial/web/rest/asab.webservice.md [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+
+We follow this up with imports, and all we will need here is our application:   
+`from myrestapi.app import CRUDApp`   
+Next, we instantiate an object of our CRUDApp class in the __main__ of the application, and we run the app:
```
Comment (ateska) [inline doc/tutorial/web/rest/asab.webservice.md:37]: Once again, `CRUDApp` can be misleading `TutorialApp` is probably better, the best would be to name it based a functionality (eg. TodoApp).

File: doc/tutorial/web/rest/asab.webservice.md [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+    
+Next, we describe the class, it inherits from the basic ASAB Application class, but we need to expand 
+it a little (see comments in the code for more details):
+    
+    
+    class CRUDApp(asab.Application):
+
+	    def __init__(self):
+		    super().__init__()
+		    # Circumvent the need for -w flag when starting the app
```
Comment (ateska) [inline doc/tutorial/web/rest/asab.webservice.md:70]: This sounds hacky ... you may rather say:  
Alternative for command-line flag `-w`

File: doc/tutorial/web/rest/asab.webservice.md [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+		    # We need to pass the CRUDService as an argument, when instantiating the class
+		    self.CRUDWebHandler = CRUDWebHandler(self, self.CRUDService)
+
+This could also be a BSPUMP.Application, enabling us to include a bspump pipeline in our microservice. 
+To do this, inherit instead from bspump.BSPumpApplication and add service, connections and pipelines as 
+usual.
+   
+#### ./myrestapi/crud/handler.py
+
+The handler is where we define the operations our microservice will be able to perform. As the
```
Comment (ateska) [inline doc/tutorial/web/rest/asab.webservice.md:93]: The handler is where HTTP Rest calls are handled and transformed into the actual (internal) service calls. From another perspective, the handler should contain only translation between REST calls and service interface. No actual 'business logic' should be here.

File: doc/tutorial/web/rest/asab.webservice.md [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+    
+Next is the CRUDWebhandler class:
+
+    class CRUDWebHandler(object):
+	def __init__(self, app, mongo_svc):
+		self.CRUDService = mongo_svc
+
+		web_app = app.WebContainer.WebApp
+		
+		web_app.router.add_put('/crud-api/{collection}', self.create) # Create endpoint url
```
Comment (ateska) [inline doc/tutorial/web/rest/asab.webservice.md:113]: `web_app.router.add_get('/crud-api/{collection}', self.read)` = Get many 

I also suggest to organize this to separate Read and Write methods.

File: doc/tutorial/web/rest/asab.webservice.md [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+	}})
+	async def create(self, request, *, json_data):
+        collection = request.match_info['collection']
+		result = await self.CRUDService.create(collection, json_data)
+		if result:
+			return asab.web.rest.json_response(request, {"result": "OK"})
+		else:
+			asab.web.rest.json_response(request, {"result": "FAIL"})
+
+	async def read(self, request):
```
Comment (ateska) [inline doc/tutorial/web/rest/asab.webservice.md:136]: `read_one`

File: doc/tutorial/web/rest/asab.webservice.md (hunk 4 above)
Comment (Plesoun) [reply doc/tutorial/web/rest/asab.webservice.md:33]: I may have outadated sources :) will correct
https://automatetheboringstuff.com/appendixb/

Comment (ateska) [reply doc/tutorial/web/rest/asab.webservice.md:33]: Well, if you have platform dependent hashbang, what is the point to use multi-platform Python?
Please use the 'universal' one.

Comment (Plesoun) [reply doc/tutorial/web/rest/asab.webservice.md:33]: Ok, will do :)

File: doc/tutorial/web/rest/asab.webservice.md (hunk 8 above)
Comment (Plesoun) [reply doc/tutorial/web/rest/asab.webservice.md:113]: "This" meaning the endpoints?

**Pull request merged**